### PR TITLE
Parallel repins

### PR DESCRIPTION
**Pull request by ro-tex**
# PULL REQUEST

## Overview

We can afford to repin multiple skylinks in parallel. Let's do that.

## Checklist

- [ ] All git commits are signed. (REQUIRED)
- [ ] All new methods or updated methods have clear docstrings.
- [ ] Testing added or updated for new methods.
- [ ] Verified if any changes impact the WebPortal Health Checks.
- [ ] Appropriate documentation updated.
- [ ] Changelog file created.

### CODE REVIEW

File: workers/scanner.go [hunk 1]
```diff
@@ -44,6 +45,12 @@ const (
 	fanoutRedundancy          = 3
 )
 
+const (
+	// scannerThreads defines the number of scanning threads which might attempt
+	// to pin an underpinned skylink.
+	scannerThreads = 5
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply workers/scanner.go:51]: Yes, good point. A F/U sounds good.

File: workers/scanner.go [hunk 2]
```diff
@@ -249,6 +284,53 @@ func (s *Scanner) threadedScanAndPin() {
...
+}
+
+// managedClearSkippedSkylinks clears the skipped skylinks.
+func (s *Scanner) managedClearSkippedSkylinks() {
+	s.mu.Lock()
+	s.skipSkylinks = []string{}
+	s.mu.Unlock()
+}
+
+// managedClearStats clears the scanning statistics.
```
Comment (MSevey) [inline workers/scanner.go:326]: Follow Up

```suggestion
// managedResetStats resets the scanning statistics.
```

File: workers/scanner.go [hunk 3]
```diff
@@ -321,21 +383,12 @@ func (s *Scanner) managedPinUnderpinnedSkylinks() {
...
-			s.mu.Unlock()
-			s.staticLogger.Infof("Time %s, runtime %s, pinned skylinks %d, skipped skylinks %d", t1.Format(conf.TimeFormat), t1.Sub(t0).String(), countPinned, numSkipped)
-		default:
-		}
-
 		skylink, sp, continueScanning, err := s.managedFindAndPinOneUnderpinnedSkylink()
 		if !sp.IsEmpty() {
-			countPinned++
-		} else {
+			// println(" ++ sp not empty", sp.Path)
```
Comment (MSevey) [inline workers/scanner.go:388]: Follow up

```suggestion
```

File: workers/scanner.go [hunk 4]
```diff
@@ -54,6 +61,10 @@ var (
 			Dev:      1 * database.MiB,
 			Testing:  1 * database.MiB,
 		}).(int)
+	// MaxRepairingSkylinks defines the maximum number of skylink which scanner
+	// is allowed to wait on (wait for their health to reach 0). When we reach
+	// this number we'll stop pinning new skylinks until we go under it again.
+	MaxRepairingSkylinks = int32(3)
```
Comment (MSevey) [inline workers/scanner.go:67]: I might of missed it, but is this used? If not, fine to remove as a follow up.

**Review by MSevey — APPROVED** [review]

File: workers/scanner.go (hunk 4 above)
Comment (ro-tex) [reply workers/scanner.go:67]: Yes, I've missed this during cleanup. I was working on a more flexible approach to this but it turned out it's not viable, so I scrapped it. It seems there are some leftovers.

File: workers/scanner.go [hunk 5]
```diff
@@ -249,6 +284,53 @@ func (s *Scanner) threadedScanAndPin() {
...
+			t1.Format(conf.TimeFormat), t1.Sub(startTime).String(), atomic.LoadUint32(&s.atomicCountPinned), len(skipped))
+		s.staticLogger.Tracef("Skipped %d skylinks: %v", len(skipped), skipped)
+	case <-s.staticTG.StopChan():
+		s.staticLogger.Trace("Stop channel closed")
+		return
+	}
+}
+
+// managedClearSkippedSkylinks clears the skipped skylinks.
+func (s *Scanner) managedClearSkippedSkylinks() {
```
[the remark this answers is not in the document]
Comment (ro-tex) [reply workers/scanner.go:320]: Good input. Will rename.

File: workers/scanner.go [hunk 6]
```diff
@@ -238,7 +245,27 @@ func (s *Scanner) threadedScanAndPin() {
...
+		var wg sync.WaitGroup
+		for i := 0; i < s.staticScannerThreads; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				s.managedPinUnderpinnedSkylinks()
+			}()
+		}
+		wg.Wait()
+		close(statsCh)
```
Comment (ChrisSchinnerl) [inline workers/scanner.go:268]: you might want to wait for the stats thread to be done as well before returning.

Goroutines which run in the background and write to files can panic if the main shutdown thread closes the file and the background thread wants to access the filehandle after that.

**Review by MSevey — APPROVED** [review]

File: workers/scanner.go (hunk 6 above)
Comment (ro-tex) [reply workers/scanner.go:268]: Good point, I'll add it to the threadgroup counter. It seems like all threaded methods should start with `defer <tg>.done()`.

**Review by ChrisSchinnerl — APPROVED** [review]

**Review by MSevey — APPROVED** [review]

**Pull request merged**